src/main/scala/com/databricks/spark/avro/AvroSaver.scala

```diff
@@ -0,0 +1,205 @@
+      case FloatType => newFieldBuilder.floatType()
+      case DoubleType => newFieldBuilder.doubleType()
+      case DecimalType() => newFieldBuilder.stringType()
+      case StringType => newFieldBuilder.stringType()
+      case BinaryType => newFieldBuilder.bytesType()
+      case BooleanType => newFieldBuilder.booleanType()
+      case TimestampType => newFieldBuilder.longType()
+
+      case ArrayType(elementType, _) =>
+        val builder = SchemaBuilder.builder()
+        val elementSchema = convertTypeToAvro(elementType, builder, structName)
+
+        newFieldBuilder.array().items(elementSchema)
+
+      case MapType(StringType, valueType, _) =>
+        val builder = SchemaBuilder.builder()
+        val valueSchema = convertTypeToAvro(valueType, builder, structName)
+
+        newFieldBuilder.map().values(valueSchema)
+
+      case StructType(_) =>
+        convertStructToAvro(dataType.asInstanceOf[StructType],
+          newFieldBuilder.record(structName))
+
+      case other => sys.error(s"Unexpected type $dataType")
+    }
+  }
+
+  private def convertStructToAvro[T](structType: StructType, schemaBuilder: RecordBuilder[T]): T = {
+    val fieldsAssembler: FieldAssembler[T] = schemaBuilder.fields()
+    structType.fields foreach { field =>
+      val newField = fieldsAssembler.name(field.name).`type`()
+
+      if (field.nullable) {
+        convertTypeToAvro(field.dataType, newField.optional(), field.name)
+      } else {
+        convertFieldTypeToAvro(field.dataType, newField, field.name).noDefault()
+      }
+    }
+    fieldsAssembler.endRecord()
+  }
+
+  private def convertValueToAvro(item: Any, dataType: DataType, structName: String): Any = {
+    if (item == null) {
+      // Since some schema fields might be nullable, item can be null. If so, just return it
+      // without further processing
+      return null
+    }
+
+    dataType match {
+      case ByteType | ShortType | IntegerType | LongType | FloatType | DoubleType | StringType |
+           BinaryType | BooleanType =>
+        item
+
+      case DecimalType() =>
+        item.toString
+
+      case TimestampType =>
+        item.asInstanceOf[Timestamp].getTime
+
+      case ArrayType(elementType, _) =>
+        val sourceArray = item.asInstanceOf[Seq[Any]]
+        val targetArray = new Array[Any](sourceArray.size)
+        var idx = 0
+
+        while (idx < sourceArray.size) {
+          targetArray(idx) = convertValueToAvro(sourceArray(idx), elementType, structName)
+          idx += 1
+        }
+
+        targetArray
+
+      case MapType(StringType, valueType, _) =>
+        val javaMap = new HashMap[String, Any]()
+        item.asInstanceOf[Map[String, Any]].foreach { case (key, value) =>
+          javaMap.put(key, convertValueToAvro(value, valueType, structName))
+        }
+        javaMap
+
+      case StructType(_) => rowToAvro(item.asInstanceOf[Row], dataType.asInstanceOf[StructType],
+        structName)._1.datum()
+    }
```

Comment:
add a `case other` like for the other guys to throw a nicer exception than MatchError